### PR TITLE
open rocksdb once

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Open RocksDB once, and fix wal-dir bug. 
@BusyJay @siddontang PTAL

### CODE REVIEW

File: src/util/rocksdb.rs [hunk 1]
```diff
@@ -46,56 +46,75 @@ pub fn open_opt(opts: Options,
...
-fn check_column_families(path: &str,
-                         needed_cfs: &[&str],
-                         cfs_opts: &[&Options])
-                         -> Result<(), String> {
+fn check_and_open(path: &str,
+                  mut db_opt: Options,
+                  cfs_opts: HashMap<&str, Options>)
+                  -> Result<DB, String> {
     // If db not exist, create it.
     if !db_exist(path) {
```
Comment (siddontang) [inline src/util/rocksdb.rs]:
```
let mut cfs_list;
if !db_exists(path) {
    db_opt.create_if_missing(true);
    cfs_list = [default];
} else {
    db_opt.create_if_missing(false);
    cfs_list = try!(DB:: list_column_families());
}
```

Then you can use the same logic, no need to handle db missing separately.

Comment (zhangjinpeng1987) [inline src/util/rocksdb.rs]: There is no same logic can be reused, on the other hand it is more clear separate the conditions db exist and not exist.

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -654,15 +654,16 @@ fn build_raftkv(config: &toml::Value,
...
     let trans = ServerTransport::new(ch);
     let path = Path::new(&cfg.storage.path).to_path_buf();
-    let opts = get_rocksdb_db_option(config);
-    let cfs_opts = vec![get_rocksdb_default_cf_option(config, total_mem),
-                        get_rocksdb_lock_cf_option(config),
-                        get_rocksdb_write_cf_option(config, total_mem),
-                        get_rocksdb_raftlog_cf_option(config, total_mem)];
+    let db_opts = get_rocksdb_db_option(config);
+    let mut cfs_opts = HashMap::default();
+    cfs_opts.insert(CF_DEFAULT, get_rocksdb_default_cf_option(config, total_mem));
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: We may need to consider save the cf opt in rocksdb later, otherwise tikv may core.

File: src/util/rocksdb.rs [hunk 3]
```diff
@@ -196,29 +205,40 @@ impl SliceTransform for FixedPrefixSliceTransform {
...
-        check_column_families(path_str, &["default"], &cfs_ref_opts[..1]).unwrap();
+        let mut cfs_opts = HashMap::default();
+        cfs_opts.insert("default", Options::new());
+        let db = check_and_open(path_str, Options::new(), cfs_opts).unwrap();
+        drop(db);
         column_families_must_eq(path_str, &["default"]);
 
         // add cf1.
-        check_column_families(path_str, &["default", "cf1"], &cfs_ref_opts).unwrap();
+        let mut cfs_opts = HashMap::default();
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: You can simplify it as

```
let cfs_opts = map!["default" => Options::new(), "cf1" => Options::new];
```

File: src/util/rocksdb.rs [hunk 4]
```diff
@@ -196,29 +205,40 @@ impl SliceTransform for FixedPrefixSliceTransform {
...
+        cfs_opts.insert("default", Options::new());
+        cfs_opts.insert("cf1", Options::new());
+        let db = check_and_open(path_str, Options::new(), cfs_opts).unwrap();
+        drop(db);
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: Why drop db explicitly?

Comment (zhangjinpeng1987) [inline src/util/rocksdb.rs]: close db prepare for next `check_and_open`.

File: src/bin/tikv-server.rs (hunk 2 above)
Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: Yes, another pr.

**Comment by siddontang** [comment]
PTAL @BusyJay

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**